lib/functions.inc.php

```diff
@@ -194,6 +194,25 @@ function check_password_strength( $password, $oldpassword, $pwd_policy_config, $
         }
     }
 
+    # is same as a custom password?
```

Comment:
Hello,

I am doing a last big review for this feature.

@markus-96 This pwd_no_reuse check seems really strange to me.

You require that the new password for "customfield1" is different from:
- standard userPassword
- plus any configured "customfieldN"

This seems, at least, very specific, and confusing for the user. It is especially confusing for the admin because `pwd_no_reuse` does not have the same meaning as in the general configuration.

Also, I think there will be use cases when the password must be the same as other custom password field. With this option, you can't tune precisely what field must be different.

For the sake of simplification, my first idea would be to remove the feature, and only keep the simple logic of verifying only the field you are trying to update. Do you have a special use case for requiring this feature @markus-96 ? Could you describe the feature to check how we could implement it?

Reply:
> Hello,
> 
> You require that the new password for "customfield1" is different from:
> 
>     * standard userPassword
> 
>     * plus any configured "customfieldN"

(for clarification: it does not check any configured customfieldN, it only checks the one with $custompwdfieldN['pwd_policy_config']['pwd_no_reuse'] set to true)

> Do you have a special use case for requiring this feature @markus-96 ?

Well, the main focus for me was to have a separate password that is stored in sambaNTPassword, and I did not want it to be the same as the main userpassword, since it uses some rather insecure hashing method. If you change the password stored in sambaNTPassword, you can directly check if it is the same as the main userpassword. But if you change the main userpassword, you have to check if it is the same as the one that is represented by the hash stored in sambaNTpassword.

Setting the same password to the customfield as to the main userpassword, for me this sounded like reusing the password. This is why I used this setting key for this. But it is right, when turning this on, it should be checked if the new password is the same as the already set one.

So, I think it should be renamed to 'must_be_unique' or something like this.

These two considerations would result in a bit more code to be written. But this feature is mandatory for me, as I already said a few times, and without it I would it would be useless for me.

Reply:
> Other point: from time to time, the custompasswordfield is not modified in the LDAP, but the userPassword instead. I'll continue the investigation on this point.

Ok, this is really strange. Do you have any hint when this happens? I will also recheck everything that could be related to this, but in the time of writing this I have no idea how this could happen. Maybe one of these:

- when calling change_password, somehow $custom_pwd_field_mode is set to false or $custom_pwd_attribute is set to "userPassword"
- with ad_mode, maybe, somehow, something is not working as intended. But I can not check this since I do not have access to an AD instance, and ad_mode is always false here.

Reply:
> (for clarification: it does not check any configured customfieldN, it only checks the one with $custompwdfieldN['pwd_policy_config']['pwd_no_reuse'] set to true)

The problem here is that it checks all attributes having $custompwdfield[$n]['pwd_policy_config']['pwd_no_reuse'] == true

For example, if I have a configuration with 3 LDAP custom password attributes : "custom1", "custom2" and "custom3", when I am updating custom1 with password secret, ssp will check if this value is matching the ones in custom1, custom2 and custom3. What would be more logic IMO is to check only custom1 value.

Reply:
> So, I think it should be renamed to 'must_be_unique' or something like this. These two considerations would result in a bit more code to be written. But this feature is mandatory for me, as I already said a few times, and without it I would it would be useless for me.

Then I think it would be better to have a dedicated parameter, like, as you proposed: `unique_across_password_fields`

Reply:
If you are ok with the new parameter, I can work on this.